jedi/evaluate/docstrings.py

```diff
@@ -45,28 +49,147 @@
 except ImportError:
     def _search_param_in_numpydocstr(docstr, param_str):
         return []
+
+    def _search_return_in_numpydocstr(docstr):
+        return []
 else:
     def _search_param_in_numpydocstr(docstr, param_str):
-        """Search `docstr` (in numpydoc format) for type(-s) of `param_str`."""
+        r"""
+        Search `docstr` (in numpydoc format) for type(-s) of `param_str`.
+
+        >>> from jedi.evaluate.docstrings import *  # NOQA
+        >>> from jedi.evaluate.docstrings import _search_param_in_numpydocstr
+        >>> docstr = (
+        ...    'Parameters\n'
+        ...    '----------\n'
+        ...    'x : ndarray\n'
+        ...    'y : int or str or list\n'
+        ...    'z : {"foo", "bar", 100500}, optional\n'
+        ... )
+        >>> _search_param_in_numpydocstr(docstr, 'x')
+        ['ndarray']
+        >>> sorted(_search_param_in_numpydocstr(docstr, 'y'))
+        ['int', 'list', 'str']
+        >>> sorted(_search_param_in_numpydocstr(docstr, 'z'))
+        ['int', 'str']
+        """
         params = NumpyDocString(docstr)._parsed_data['Parameters']
         for p_name, p_type, p_descr in params:
             if p_name == param_str:
                 m = re.match('([^,]+(,[^,]+)*?)(,[ ]*optional)?$', p_type)
                 if m:
                     p_type = m.group(1)
-
-                if p_type.startswith('{'):
-                    types = set(type(x).__name__ for x in literal_eval(p_type))
-                    return list(types)
-                else:
-                    return [p_type]
+                return _expand_typestr(p_type)
```

Comment:
Moved this logic to a new function so it can be used by google style docstrings as well.